File: android/app/src/main/java/com/brekeke/phonedev/call_history/CallService.java
package com.brekeke.phonedev.call_history;

import android.content.Intent;
import android.net.Uri;
import android.telecom.Call;
import android.telecom.InCallService;
import android.telecom.TelecomManager;

public class CallService extends InCallService {
  @Override
  public void onCallAdded(Call call) {
    super.onCallAdded(call);
    call.registerCallback(callCallback);
    // we can start our call Activity here
    try {
      if (call.getDetails().getHandlePresentation() == TelecomManager.PRESENTATION_ALLOWED) {
        // this is an outgoing call
        String number = call.getDetails().getHandle().getSchemeSpecificPart();
        if (number == null) {
          return;
        }
        Uri phoneNumberUri = Uri.parse("tel:" + number);
        Intent intent = new Intent(Intent.ACTION_VIEW, phoneNumberUri);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.addCategory(Intent.CATEGORY_BROWSABLE);
        startActivity(intent);
      } else {
        // TODO:incoming call
      }
    } catch (Exception e) {

    }
  }

  @Override
  public void onCallRemoved(Call call) {
    super.onCallRemoved(call);
    call.unregisterCallback(callCallback);
  }

  private Call.Callback callCallback =
      new Call.Callback() {
        @Override
        public void onStateChanged(Call call, int state) {}
      };
}

File: apps/brekekephone/app/android/app/src/main/java/com/brekeke/phonedev/call_history/CallService.kt
package com.brekeke.phonedev.call_history

import android.content.Intent
import android.net.Uri
import android.telecom.Call
import android.telecom.InCallService
import android.telecom.TelecomManager

class CallService : InCallService() {
  override fun onCallAdded(call: Call) {
    super.onCallAdded(call)
    call.registerCallback(callCallback)
    try {
      if (call.details.handlePresentation == TelecomManager.PRESENTATION_ALLOWED) {
        val number = call.details.handle?.schemeSpecificPart ?: return
        val phoneNumberUri = Uri.parse("tel:$number")
        val intent = Intent(Intent.ACTION_VIEW, phoneNumberUri)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.addCategory(Intent.CATEGORY_BROWSABLE)
        startActivity(intent)
      } else {
        // TODO: incoming call
      }
    } catch (_: Exception) {}
  }

  override fun onCallRemoved(call: Call) {
    super.onCallRemoved(call)
    call.unregisterCallback(callCallback)
  }

  private val callCallback =
      object : Call.Callback() {
        override fun onStateChanged(call: Call, state: Int) {}
      }
}
